Mathlib/NumberTheory/DirichletCharacter/GaussSum.lean

```diff
@@ -0,0 +1,77 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+import Mathlib.NumberTheory.DirichletCharacter.Basic
+import Mathlib.NumberTheory.LegendreSymbol.GaussSum
+
+/-!
+# Gauss sums for Dirichlet characters
+-/
+variable {N : ℕ+} {R : Type*} [CommRing R] (e : AddChar (ZMod N) R)
+
+open AddChar
+
+lemma gaussSum_aux_of_mulShift (χ : DirichletCharacter R N) {d : ℕ}
+    (hd : d ∣ N) (he : e.mulShift d = 1) {u : (ZMod N)ˣ} (hu : ZMod.unitsMap hd u = 1) :
+    χ u * gaussSum χ e = gaussSum χ e := by
+  suffices e.mulShift u = e by conv_lhs => rw [← this, gaussSum_mulShift]
+  rw [(by ring : u.val = (u - 1) + 1), ← mulShift_mul, mulShift_one, mul_left_eq_self]
+  rsuffices ⟨a, ha⟩ : (d : ℤ) ∣ (u.val.val - 1 : ℤ)
+  · have : u.val - 1 = ↑(u.val.val - 1 : ℤ) := by simp only [ZMod.natCast_val, Int.cast_sub,
+      ZMod.intCast_cast, ZMod.cast_id', id_eq, Int.cast_one]
+    rw [this, ha]
+    ext1 y
+    simpa only [Int.cast_mul, Int.cast_natCast, mulShift_apply, mul_assoc, one_apply]
+      using DFunLike.ext_iff.mp he (a * y)
+  rw [← Units.eq_iff, Units.val_one, ZMod.unitsMap_def, Units.coe_map] at hu
+  have : ZMod.castHom hd (ZMod d) u.val = ((u.val.val : ℤ) : ZMod d) := by simp
+  rwa [MonoidHom.coe_coe, this, ← Int.cast_one, eq_comm,
+    ZMod.intCast_eq_intCast_iff_dvd_sub] at hu
+
+/-- If `gaussSum χ e ≠ 0`, and `d` is such that `e.mulShift d = 1`, then `χ` must factor through
+`d`. (This will be used to show that Gauss sums vanish when `χ` is primitive and `e` is not.)-/
+lemma factorsThrough_of_gaussSum_ne_zero [IsDomain R] {χ : DirichletCharacter R N} {d : ℕ}
+    (hd : d ∣ N) (he : e.mulShift d = 1) (h_ne : gaussSum χ e ≠ 0) :
+    χ.FactorsThrough d := by
+  rw [DirichletCharacter.factorsThrough_iff_ker_unitsMap hd]
+  intro u hu
+  rw [MonoidHom.mem_ker, ← Units.eq_iff, MulChar.coe_toUnitHom]
+  simpa only [Units.val_one, ne_eq, h_ne, not_false_eq_true, mul_eq_right₀] using
+    gaussSum_aux_of_mulShift e χ hd he hu
+
+/-- If `χ` is primitive, but `e` is not, then `gaussSum χ e = 0`. -/
+lemma gaussSum_eq_zero_of_isPrimitive_of_not_isPrimitive [IsDomain R]
+    {χ : DirichletCharacter R N} (hχ : χ.isPrimitive) (he : ¬e.IsPrimitive) :
+    gaussSum χ e = 0 := by
+  contrapose! hχ
+  rcases e.exists_divisor_of_not_isPrimitive he with ⟨d, hd₁, hd₂, hed⟩
+  have : χ.conductor ≤ d := Nat.sInf_le <| factorsThrough_of_gaussSum_ne_zero e hd₁ hed hχ
+  exact (this.trans_lt hd₂).ne
+
+/-- If `χ` is a primitive character, then the function `a ↦ gaussSum χ (e.mulShift a)`, for any
+fixed additive character `e`, is a constant multiple of `χ⁻¹`. -/
+lemma gaussSum_mulShift_of_isPrimitive [IsDomain R] {χ : DirichletCharacter R N}
+    (hχ : χ.isPrimitive) (a : ZMod N) :
+    gaussSum χ (e.mulShift a) = χ⁻¹ a * gaussSum χ e := by
+  by_cases ha : IsUnit a
+  · conv_rhs => rw [← gaussSum_mulShift χ e ha.unit]
+    have : IsUnit (χ a) := by rw [← ha.unit_spec, ← MulChar.coe_toUnitHom]; apply Units.isUnit
```

Comment:
Maybe add
```lean
lemma MulChar.map_unit (χ : MulChar R R') {a : R} (ha : IsUnit a) : IsUnit (χ a)
```
(with suitable `R` and `R'`) in `Mathlib.NumberTheory.LegendreSymbol.MulCharacter`? 
It looks like a generally useful statement.

Reply:
In the course of attempting to add it, I realised someone had already done so – it's `IsUnit.map`.